### PR TITLE
Sophiex/kerem/pr/transformer head

### LINKED ISSUE #1587: Implement transformer based predictors for JEPA
### Describe the task. It can be a feature, documentation, etc.

We need to implement a transformer based predictor for JEPA, following the literature (see https://www.overleaf.com/project/68dd158b1e327e168d9ceebd)

We will use the existing attention blocks and MLP layers for this. This resembles the forecasting engine architecture. A quick look revelas that it is not very straight forward to refactor these two to a single class, so first we will implement it as a seperate class.

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### Area

- [ ] datasets, data readers, data preparation and transfer
- [x] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation

### PR DESCRIPTION
**Pull request by sophie-xhonneux**
## Description

See PR #1590

## Issue Number

Closes #1587 

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [ ] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: config/config_physical_jepa.yml [hunk 1]
```diff
@@ -64,6 +64,12 @@ fe_impute_latent_noise_std: 0.0  # 1e-4
 # currently fixed to 1.0 (due to limitations with flex_attention and triton)
 forecast_att_dense_rate: 1.0
 
+sslpred_num_blocks: 12
```
Comment (clessig) [inline config/config_physical_jepa.yml:67]: I don't think these params should go into the model block but part of the SSL loss term. This here is really also JEPA specific.

File: src/weathergen/model/engines.py [hunk 2]
```diff
@@ -835,15 +835,87 @@ class LatentState:
...
+
+        self.name = name
+        self.cf = cf
+        self.class_token = class_token
+        self.patch_token = patch_token
+
+        self.pred_blocks = nn.ModuleList()
+
+        # first map to intermediate_dim to introduce a bottleneck
+        self.pred_blocks.append(nn.Linear(in_dim, intermediate_dim, bias=False))
```
Comment (clessig) [inline src/weathergen/model/engines.py:859]: We should call this `blocks` in all modules.

File: src/weathergen/model/engines.py [hunk 3]
```diff
@@ -835,15 +835,87 @@ class LatentState:
...
+            )
+
+        # finally map from intermediate_dim to the out_dim
+        self.pred_blocks.append(nn.Linear(intermediate_dim, out_dim, bias=False))
+
+    def forward(self, x: LatentState):
+        # we concatenate the patch and class tokens to process them together
+        # We concatenate in the token dimension [Batch, Tokens, Dim]
+        patch_class_tokens = []
+        if self.class_token:
```
Comment (clessig) [inline src/weathergen/model/engines.py:896]: We should use x.class_token and x.patch_token here.

File: src/weathergen/model/engines.py [hunk 4]
```diff
@@ -835,15 +835,87 @@ class LatentState:
...
+            patch_class_tokens.append(x.class_token)
+        if self.patch_token:
+            patch_class_tokens.append(x.patch_tokens)
+        patch_class_tokens = torch.cat(patch_class_tokens, dim=1)
+
+        for _b_idx, block in enumerate(self.pred_blocks):
+            if isinstance(block, torch.nn.modules.normalization.LayerNorm):
+                patch_class_tokens = block(patch_class_tokens)
+            else:
+                patch_class_tokens = checkpoint(block, patch_class_tokens, use_reentrant=False)
```
Comment (clessig) [inline src/weathergen/model/engines.py:906]: The checkpoint should be on a coarser level if possible.

File: src/weathergen/model/engines.py [hunk 5]
```diff
@@ -835,15 +835,87 @@ class LatentState:
...
+class LatentPredictionHeadMLP(nn.Module):
+    def __init__(self, name, in_dim, out_dim, num_layers, hidden_factor, class_token: bool, patch_token: bool):
         super().__init__()
 
         self.name = name
         self.class_token = class_token
         self.patch_token = patch_token
         # For now this is a Linear Layer TBD what this architecture should be
-        self.layer = nn.Linear(in_dim, out_dim, bias=False)
+        self.layer = MLP(in_dim, out_dim, num_layers, hidden_factor)
```
Comment (clessig) [inline src/weathergen/model/engines.py:918]: `self.layer` -> `self.blocks`

File: src/weathergen/model/model.py [hunk 6]
```diff
@@ -434,6 +437,32 @@ def create(self) -> "Model":
...
+
+        def _create_latent_pred_head(
+            global_config, name, loss_conf, in_dim, class_token, patch_token
+        ):
+            global_config = OmegaConf.merge(global_config, loss_conf)
+            if loss_conf["head"] == "mlp":
+                return LatentPredictionHeadMLP(
+                    name,
+                    global_config.ae_global_dim_embed,
+                    loss_conf["out_dim"],
```
Comment (clessig) [inline src/weathergen/model/model.py:449]: This should be nested parameters for head. So the config should be something like:

head : {
  type: mlp,
  out_dim: ...
}

File: src/weathergen/model/engines.py [hunk 7]
```diff
@@ -835,15 +835,87 @@ class LatentState:
     z_pre_norm: torch.Tensor
 
 
-class LatentPredictionHead(nn.Module):
-    def __init__(self, name, in_dim, out_dim, class_token: bool, patch_token: bool):
+class LatentPredictionHeadTransformer(nn.Module):
```
Comment (clessig) [inline src/weathergen/model/engines.py:838]: This is really a vanilla self-attention transformer with the class / patch_token preprocessing that one could wrap easily in a very thing class. But we can potentially do this later

File: src/weathergen/model/model.py [hunk 8]
```diff
@@ -434,6 +437,32 @@ def create(self) -> "Model":
             for _, v in cf.training_config.losses.items()
             if v.type == "LossLatentSSLStudentTeacher"
         ]
+
+        def _create_latent_pred_head(
+            global_config, name, loss_conf, in_dim, class_token, patch_token
+        ):
+            global_config = OmegaConf.merge(global_config, loss_conf)
```
Comment (clessig) [inline src/weathergen/model/model.py:444]: Use `merge_configs()` in config.py

Comment (clessig) [reply src/weathergen/model/model.py:444]: But I think we should not merge. In the (near) future, we have model_cfg and loss_cfg, and we should use them separately and get the different params from the respective config. This is clearer.

File: src/weathergen/model/model.py [hunk 9]
```diff
@@ -434,6 +437,32 @@ def create(self) -> "Model":
             for _, v in cf.training_config.losses.items()
             if v.type == "LossLatentSSLStudentTeacher"
         ]
+
+        def _create_latent_pred_head(
+            global_config, name, loss_conf, in_dim, class_token, patch_token
```
Comment (clessig) [inline src/weathergen/model/model.py:442]: Please use a consistent naming instead of `global_config` and `loss_conf`. In Trainer we use `*_cfg` and I think we should consistently switch to this.

File: src/weathergen/model/model_interface.py [hunk 10]
```diff
@@ -114,6 +114,10 @@ def init_model_and_shard(
             if isinstance(module, modules_to_shard):
                 fully_shard(module, **fsdp_kwargs)
 
+        for module in model.latent_heads.modules():
```
Comment (clessig) [inline src/weathergen/model/model_interface.py:117]: We need to find a way that we don't have to specify it manually for each new block/module. base class?

File: src/weathergen/train/trainer.py [hunk 11]
```diff
@@ -147,6 +147,8 @@ def get_target_aux_calculators(self, mode_cfg):
         batch_size = get_batch_size_from_config(mode_cfg)
 
         # get target_aux calculators for different loss terms
+        # del self.cf.training_config.losses["student-teacher"]["loss_fcts"]["JEPA"]
```
Comment (clessig) [inline src/weathergen/train/trainer.py:150]: Remove

File: src/weathergen/train/trainer.py [hunk 12]
```diff
@@ -248,6 +250,7 @@ def run(self, cf, devices, run_id_contd=None, mini_epoch_contd=None):
             self.validate_with_ema = validate_with_ema_cfg.get("enabled", True)
         else:
             self.validate_with_ema = False
+        self.validate_with_ema = False
```
Comment (clessig) [inline src/weathergen/train/trainer.py:253]: Remove

File: src/weathergen/train/trainer.py [hunk 13]
```diff
@@ -269,7 +272,7 @@ def run(self, cf, devices, run_id_contd=None, mini_epoch_contd=None):
 
         # get target_aux calculators for different loss terms
         self.target_and_aux_calculators = self.get_target_aux_calculators(self.training_cfg)
-        self.validate_with_ema_cfg = self.get_target_aux_calculators(self.validation_cfg)
+        # self.validate_with_ema_cfg = self.get_target_aux_calculators(self.validation_cfg)
```
Comment (clessig) [inline src/weathergen/train/trainer.py:275]: Re-enable

File: src/weathergen/train/trainer.py [hunk 14]
```diff
@@ -398,8 +401,10 @@ def train(self, mini_epoch):
         # training loop
         self.t_start = time.time()
         for bidx, batch in enumerate(dataset_iter):
+            print("Happy to be here")
```
Comment (clessig) [inline src/weathergen/train/trainer.py:404]: Remove

File: src/weathergen/train/trainer.py [hunk 15]
```diff
@@ -398,8 +401,10 @@ def train(self, mini_epoch):
...
             batch.to_device(self.device)
 
+            print("Batch to device")
```
Comment (clessig) [inline src/weathergen/train/trainer.py:407]: Remove

File: src/weathergen/train/trainer.py [hunk 16]
```diff
@@ -411,6 +416,7 @@ def train(self, mini_epoch):
                     self.training_cfg.window_offset_prediction,
                 )
 
+                print("Model predictions")
```
Comment (clessig) [inline src/weathergen/train/trainer.py:419]: Remove

File: src/weathergen/train/trainer.py [hunk 17]
```diff
@@ -423,12 +429,14 @@ def train(self, mini_epoch):
                         self.model,
                         self.training_cfg.window_offset_prediction,
                     )
+                print("target predictions")
```
Comment (clessig) [inline src/weathergen/train/trainer.py:432]: Remove

File: src/weathergen/train/trainer.py [hunk 18]
```diff
@@ -423,12 +429,14 @@ def train(self, mini_epoch):
...
 
             loss = self.loss_calculator.compute_loss(
                 preds=preds,
                 targets_and_aux=targets_and_auxs,
                 metadata=extract_batch_metadata(batch),
             )
+            print("loss calcuclation")
```
Comment (clessig) [inline src/weathergen/train/trainer.py:439]: Remove

File: src/weathergen/train/trainer_base.py [hunk 19]
```diff
@@ -52,15 +52,16 @@ def init_torch(use_cuda=True, num_accs_per_task=1, multiprocessing_method="fork"
         if not use_cuda:
             return torch.device("cpu")
 
-        local_id_node = os.environ.get("SLURM_LOCALID", "-1")
+        local_id_node = os.environ.get("LOCAL_RANK", os.environ.get("SLURM_LOCALID", "-1"))
```
Comment (clessig) [inline src/weathergen/train/trainer_base.py:55]: This should be `torch.distributed.get_local_rank()`, unless someone can explain why it's not suitable here.

**Review by clessig — COMMENTED** [review]
`trainer.py` and `trainer_base.py` need to be cleaned up, please.

File: config/config_physical_jepa.yml (hunk 1 above)
Comment (kctezcan) [reply config/config_physical_jepa.yml:67]: Sophie took care of this

File: src/weathergen/model/engines.py (hunk 2 above)
Comment (kctezcan) [reply src/weathergen/model/engines.py:859]: done

File: src/weathergen/train/trainer_base.py (hunk 19 above)
Comment (kctezcan) [reply src/weathergen/train/trainer_base.py:55]: I am not sure why this even changed...

Using `dist.get_node_local_rank (fallback_rank=-1)` instead

File: src/weathergen/model/engines.py (hunk 3 above)
Comment (kctezcan) [reply src/weathergen/model/engines.py:896]: I renamed these to
```python
self.use_class_token = use_class_token
self.use_patch_token = use_patch_token
``` 
clarify that these are boolians. Now t looks like in the `forward()`
```python
if self.use_class_token:
   patch_class_tokens.append(x.class_token)
if self.use_patch_token:
   patch_class_tokens.append(x.patch_tokens)
patch_class_tokens = torch.cat(patch_class_tokens, dim=1)
```

File: src/weathergen/model/engines.py (hunk 4 above)
Comment (kctezcan) [reply src/weathergen/model/engines.py:906]: I did not understand this comment :)

File: src/weathergen/model/engines.py (hunk 5 above)
Comment (kctezcan) [reply src/weathergen/model/engines.py:918]: done

File: src/weathergen/train/trainer.py (hunk 16 above)
Comment (kctezcan) [reply src/weathergen/train/trainer.py:419]: done

File: src/weathergen/train/trainer.py (hunk 17 above)
Comment (kctezcan) [reply src/weathergen/train/trainer.py:432]: done

File: src/weathergen/train/trainer.py (hunk 18 above)
Comment (kctezcan) [reply src/weathergen/train/trainer.py:439]: done. 

removed more print() statements

File: src/weathergen/model/engines.py (hunk 4 above)
Comment (clessig) [reply src/weathergen/model/engines.py:906]: The checkpoint should wrap the call to this forward function, not be in the forward function

Comment (clessig) [reply src/weathergen/model/engines.py:906]: > I did not understand this comment :)


The checkpoint should wrap the call to this forward function, not be in the forward function

File: src/weathergen/model/engines.py (hunk 3 above)
Comment (clessig) [reply src/weathergen/model/engines.py:896]: Better although I think it's still duplicate and using `x.class_token` and `x.patch_token` is more robust.
